torch/csrc/cuda/memory_snapshot.cpp

```diff
@@ -0,0 +1,125 @@
+#include <c10/cuda/CUDACachingAllocator.h>
+#include <torch/csrc/cuda/memory_snapshot.h>
+#include <torch/csrc/jit/serialization/pickler.h>
+namespace torch {
+namespace cuda {
+
+using c10::Dict;
+using c10::IValue;
+using torch::jit::Pickler;
+
+using c10::cuda::CUDACachingAllocator::BlockInfo;
+using c10::cuda::CUDACachingAllocator::History;
+using c10::cuda::CUDACachingAllocator::SegmentInfo;
+
+namespace {
+std::unique_ptr<c10::cuda::CUDACachingAllocator::Context> blank_context() {
+  // in the future the C++-only version of context gathering could include C++
+  // or torchscript frames.
+  return std::make_unique<c10::cuda::CUDACachingAllocator::Context>();
+}
+std::vector<char> write_pickle(const IValue& v) {
+  std::vector<char> result;
+  {
+    auto writer = [&](const char* data, size_t size) {
+      result.insert(result.end(), data, data + size);
+    };
+    Pickler pickler(writer, nullptr, nullptr, nullptr, nullptr, false);
+    pickler.protocol();
+    pickler.pushIValue(v);
+    pickler.stop();
+  }
+  return result;
+}
+Dict<IValue, IValue> new_dict() {
+  return Dict<IValue, IValue>(c10::AnyType::get(), c10::AnyType::get());
+}
+c10::List<IValue> new_list() {
+  return List<IValue>(c10::AnyType::get());
+}
+} // namespace
+void _record_memory_history(bool enabled) {
+  c10::cuda::CUDACachingAllocator::setContextRecorder(
+      enabled ? blank_context : nullptr);
+}
+
+std::vector<char> _memory_snapshot_pickled() {
```

Comment:
going into a vector<char> is kinda weird but OK

Reply:
A std::string instead? I don't want it to be a file because these things might need to be written via API calls rather than directly to a file.

Reply:
going to a file stream would be the normal thing, but I think this is fine, we can reevaluate if this ends up being too big to store in memory (it shouldn't ever, tbh)